### ISSUE TITLE
Lazy/cached count

### ISSUE THREAD

**Issue by ryan-williams** [hammerlab/magic-rdds#2]
An implicitly-available field that caches the count of a given RDD so inadvertently calling more than once (e.g. in a notebook) can be smarter.

Bonus: be smart about UnionRDDs / integrate with #1.

**Comment by sadikovi** [comment]
Will this work?

``` scala

import java.util.HashMap
import org.apache.spark.rdd.RDD

object CachedCount {
  private val map: HashMap[Int, Long] = new HashMap()

  def cachedCount(rdd: RDD[_]): Long = {
    if (!map.containsKey(rdd.id)) {
      map.put(rdd.id, rdd.count)
    }
    map.get(rdd.id)
  }
}

implicit class RichRDDFunctions(rdd: RDD[_]) {
  def size(): Long = {
    CachedCount.cachedCount(rdd)
  }
}
```

Example:

``` scala
val rdd = sc.parallelize(0 until 10)
rdd.size()
// 16/08/15 22:26:40 INFO TaskSetManager: Finished task 3.0 in stage 0.0 (TID 3) in 46 ms on localhost (4/4)
// 16/08/15 22:26:40 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool 
// 16/08/15 22:26:40 INFO DAGScheduler: ResultStage 0 (count at <console>:21) finished in 0.106 s
// 16/08/15 22:26:40 INFO DAGScheduler: Job 0 finished: count at <console>:21, took 0.322182 s
// res0: Long = 10

rdd.size()
// res1: Long = 10
```

**Comment by ryan-williams** [comment]
That looks good @sadikovi, thanks! I'd use a scala `mutable.HashMap` and `getOrElseUpdate`, personally.

Cool trick with the `implicit class`, as well, though [this](http://docs.scala-lang.org/overviews/core/implicit-classes.html#restrictions) implies that it'd need to be inside the `object CachedCount` perhaps?

**Comment by sadikovi** [comment]
As an option, yes, you will probably have to import everything inside object to work. I would define special package object for all implicits, like this:

``` Scala
package object implicits {
  implicit class Test(...) {
    ...
  }
}

import implicits._
```

But it can be done with function, similar to other examples in repository.

**Comment by sadikovi** [comment]
I did not get that part on union rdds. Can you give an example? By the way, I can create PR on this change, if you want.

**Comment by ryan-williams** [comment]
Sure, a PR for this is welcome if you want to do it!

re: union RDDs, I mean that if you ever have code that does something like this:

``` scala
val count1 = rdd1.count
val count2 = rdd2.count
```

we could instead add the ability to do something like:

``` scala
val (count1, count2) = (rdd1, rdd2).multiCount
```

and this would compute both counts in one job, by making a UnionRDD (`rdd1 ++ rdd2`), counting the elems in all partitions, and then summing up the counts for each of `rdd1` and `rdd2` from those… likewise for >2 RDDs.

lmk if that doesn't make sense

**Comment by sadikovi** [comment]
Yes, I understand now, thanks, will try adding it.

**Comment by ryan-williams** [comment]
Cool, the union-rdds / multi-count bit can totally be done separately, too, if you prefer!

**Comment by ryan-williams** [comment]
Fixed by #13
